rasa/cli/export.py

```diff
@@ -0,0 +1,508 @@
+    time range.
+
+    Args:
+        events: List of serialized events to be sorted and selected from.
+        minimum_timestamp: Minimum timestamp of events that are published. If `None`,
+            apply no such constraint.
+        maximum_timestamp: Maximum timestamp of events that are published. If `None`,
+            apply no such constraint.
+
+    Returns:
+        List of serialized and sorted (by timestamp) events within the requested time
+        range.
+
+    """
+    cli_utils.print_info(
+        f"Sorting and selecting from {len(events)} total events found."
+    )
+    # sort the events by timestamp just in case they're not sorted already
+    events = sorted(events, key=lambda x: x["timestamp"])
+
+    # drop events failing minimum timestamp requirement
+    if minimum_timestamp is not None:
+        events = itertools.dropwhile(
+            lambda x: x["timestamp"] < minimum_timestamp, events
+        )
+
+    # select events passing maximum timestamp requirement
+    if maximum_timestamp is not None:
+        events = itertools.takewhile(
+            lambda x: x["timestamp"] < maximum_timestamp, events
+        )
+
+    events = list(events)
+    if not events:
+        cli_utils.print_error_and_exit(
+            "Could not find any events within requested time range. Exiting."
+        )
+
+    return events
+
+
+def _fetch_events_within_time_range(
+    tracker_store: TrackerStore,
+    minimum_timestamp: Optional[float] = None,
+    maximum_timestamp: Optional[float] = None,
+    conversation_ids: Optional[List[Text]] = None,
+) -> List[Dict[Text, Any]]:
+    """Fetch all events for `conversation_ids` within the supplied time range.
+
+    Args:
+        tracker_store: Tracker store to source events from.
+        minimum_timestamp: Minimum timestamp of events that are published. If `None`,
+            apply no such constraint.
+        maximum_timestamp: Maximum timestamp of events that are published. If `None`,
+            apply no such constraint.
+         conversation_ids: List of conversation IDs selected for publishing.
+
+    Returns:
+        List of serialized events with added `sender_id` field.
+
+    """
+    cli_utils.print_info(
+        f"Fetching events for {len(conversation_ids)} conversation IDs:"
+    )
+
+    events = []
+
+    for conversation_id in tqdm(conversation_ids, "conversation IDs"):
+        tracker = tracker_store.retrieve(conversation_id)
+        if not tracker:
+            logger.info(
+                f"Could not retrieve tracker for conversation ID "
+                f"'{conversation_id}'. Skipping."
+            )
+            continue
+
+        _events = tracker.current_state(EventVerbosity.ALL)["events"]
+
+        if not _events:
+            logger.info(
+                f"No events to migrate for conversation ID '{conversation_id}'."
+            )
+
+        # the conversation IDs are needed in the event publishing
+        for event in _events:
```

Comment:
I see some similarity to `EventBroker.stream_events` here.
How about add a method `stream_events_in_time_span(start, end)` to `EventBroker` which does the publishing? This would publish the events one conversation id at a time which could also be better  than loading all events into memory.

Reply:
I prefer this approach because it allows us to continue when something goes wrong using only the `minimum_timestamp` and `maximum_timestamp` flags. Doing the migration per conversation ID makes it way more complicated

Reply:
mhm, ok. See what you mean. Still not exactly liking the duplication of the logic to prepare an event for publishing 🤔

Reply:
adding `sender_id` before publishing happens twice in the code base, which IMO is too early for abstraction :D